src/uitext.ml

```diff
@@ -1248,7 +1248,89 @@ let setupSafeStop () =
     if ok then stopPipe := Some (Lwt_unix.pipe_in ~cloexec:true ()))
   end
 
+(*** Requesting safe termination via stdin ***)
+
+let nonEofStopCond s =
+  String.contains s '\004' (* ^D *)
+
+let selectStdin () =
+  match Unix.select [Unix.stdin] [] [] 0. with
+  | ([r1], _, _) when r1 ==(*phys*) Unix.stdin -> true
+  | _ -> false
+
+let rec readStdinNoIntr b l =
+  try Unix.read Unix.stdin b 0 l
+  with Unix.Unix_error (EINTR, _, _) -> readStdinNoIntr b l
+
+external fdIsReadable : Unix.file_descr -> bool = "unsn_fd_readable"
+
+let stdinIsatty = Unix.isatty Unix.stdin
+
+let stdinOkForStopReq =
+  let rec notEOF st =
+    if stdinIsatty then true else
+    match selectStdin () with
+    | true ->
+        let l = 32 in
+        let b = Bytes.create l in
+        begin match readStdinNoIntr b l with
+        | 0 -> false
+        | n ->
+            if nonEofStopCond (Bytes.sub_string b 0 n) then requestSafeStop ();
+            true
+        | exception Unix.Unix_error _ -> true
+        end
+    | false -> true
+    | exception Unix.Unix_error (EINTR, _, _) -> notEOF st
+    | exception Unix.Unix_error (EBADF, _, _) -> true
+  in
+  match Unix.LargeFile.fstat Unix.stdin with
+  | { st_kind = S_REG; _ } -> false
+  | st -> begin try fdIsReadable Unix.stdin with Unix.Unix_error _ -> true end && notEOF st
+  | exception Unix.Unix_error _ -> false
+
```

Comment:
Is this the /dev/null check, that if you get EOF on the very first read, then it isn't used?  I didn't figure that out from the manual, and it provides a race condition where termination might be signalled before unison does the first read.

Reply:
You are correct. Yes, this is the check and yes, there is a race condition. However, I decided to accept this race condition because it is extremely unlikely to occur with "normal" usage. I don't know of another way of checking specifically for /dev/null and distinguishing it from something else that just happens to already return EOF.

Reply:
That may or may not be ok,  but the documentation did not explain the behavior or caution, and did not hint at a race.